asf_smart_meter_exploration/pipeline/data_aggregation.py

```diff
@@ -0,0 +1,177 @@
+    )
+
+    # Calculate means for each half-hour and day type pair
+    meter_data_daytypes = (
+        meter_data.groupby(["weekend_or_bank_holiday", "time"])
+        .mean(numeric_only=True)
+        .dropna(axis=1)
+        .T
+    )
+
+    if normalise:
+        meter_data_daytypes_norm = meter_data_daytypes.div(
+            meter_data_daytypes.groupby(level=0, axis=1).transform("sum"), axis=0
+        ).dropna(axis=0)
+        return meter_data_daytypes_norm
+    else:
+        return meter_data_daytypes
+
+
+def get_daytype_diff(type="diff"):
+    """For each household, get difference or ratio between weekend and weekday usage
+    for each half-hour of the day. ("Weekend" also includes bank holidays.)
+
+    Args:
+        type (str, optional): Whether to calculate difference ("diff") or ratio ("ratio").
+        "diff" is weekend - weekday, "ratio" is weekend / weekday.
+        Defaults to "diff".
+
+    Returns:
+        pd.DataFrame: Average differences/ratios between usage on weekends and weekdays.
+    """
+
+    meter_data_daytypes = get_average_usage_daytypes()
+
+    if type == "diff":
+        # True = "weekend or bank holiday"
+        return (meter_data_daytypes[True] - meter_data_daytypes[False]).dropna(axis=0)
+    elif type == "ratio":
+        return (
+            (meter_data_daytypes[True] / meter_data_daytypes[False])
+            .replace(np.inf, np.nan)
+            .dropna(axis=0)
+        )
+    else:
+        raise ValueError("Type must be one of 'diff' or 'ratio'.")
+
+
+def get_season_diff(meter_data=meter_data, season_1="winter", season_2="summer"):
+    """For each household, get difference between usage in two specified seasons for each half-hour.
+    Calculation performed is (mean in season_1) - (mean in season_2).
+
+    Args:
+        meter_data (_type_, optional): _description_. Defaults to meter_data.
+        season_1 (str, optional): Season, i.e. "winter", "spring", "summer" or "autumn".
+            Defaults to "winter".
+        season_2 (str, optional): Season to subtract. Can also pass "spring and autumn" to get
+            mean over both seasons. Defaults to "summer".
+
+    Returns:
+        pd.DataFrame: Average differences between usage in the two seasons.
+    """
+
+    season_dict = {
+        "winter": 0,
+        "spring": 1,
+        "summer": 2,
+        "autumn": 3,
+    }
+
+    meter_data["season"] = (
+        meter_data.tstp.dt.month // 3 % 4
+    )  # quick way of getting season number
+    seasonal_aves = (
+        meter_data.groupby(["season", "time"]).mean(numeric_only=True).dropna(axis=1)
+    )
+
+    if season_2 != "spring and autumn":
+        seasonal_diff = (
+            seasonal_aves.loc[season_dict[season_1]]
+            - seasonal_aves.loc[season_dict[season_2]]
+        ).T
+    else:
+        spring_autumn_ave = (seasonal_aves.loc[1] + seasonal_aves.loc[3]) / 2
+        seasonal_diff = (seasonal_aves.loc[season_dict[season_1]] - spring_autumn_ave).T
```

Comment:
Mainly because this was the only combination I was interested in, so didn't want to spend the time making it more general! The reason being that seasonal variation is mainly due to heating/cooling, so we would expect spring and autumn to be pretty similar and it makes more sense to combine them than any other two seasons. I might not generalise this now but definitely something that could be done later